File: app/models/Switch.scala
package models

import play.api.libs.json.{Json, OFormat}

case class Switch(
  name: String,
  ip: String,
  mac: String,
  snmpCommunity: String,
  revision: Option[String],
  serial: Option[String],
  portsNumber: Option[Int],
  buildShortName: Option[String],
  floorNumber: Option[Int],
  positionTop: Option[Float],
  positionLeft: Option[Float],
  upSwitchName: Option[String],
  upSwitchMAC: Option[String],
  upLink: Option[String]
) {
  override def equals(that: Any): Boolean = false
}

object Switch {
  def tupled: (
    (
      String,
      String,
      String,
      String,
      Option[String],
      Option[String],
      Option[Int],
      Option[String],
      Option[Int],
      Option[Int],
      Option[Int],
      Option[String],
      Option[String],
      Option[String]
    )
  ) => Switch = (Switch.apply _).tupled
  implicit val switchFormat: OFormat[Switch] = Json.format[Switch]
}

File: ui/src/types/switch.ts
export type SwitchRequest = SwitchResponse & {
  snmpCommunity: string;
  retrieveFromNetData: boolean;
  retrieveUpLinkFromSeens: boolean;
  retrieveTechDataFromSNMP: boolean;
  ipResolveMethod: string;
};

export type SwitchResponse = {
  name: string;
  ip: string;
  mac: string;
  revision: string;
  serial: string;
  buildShortName: string;
  floorNumber: number;
  positionTop: number;
  positionLeft: number;
  upSwitchName: string;
  upLink: string;
};

export type SwitchResult = {
  sw: SwitchResponse;
  seen: boolean;
  snmp: boolean;
};

export type SavePositionRequest = {
  top: number;
  left: number;
};
